### ISSUE TITLE
`requires_grad_` in `no_grad` context returns incorrect value with tensor subclasses

### ISSUE THREAD

**Issue by jph00** [pytorch/pytorch#50219]
## 🐛 Bug

When `requires_grad_` is called in a `no_grad` context on a `Tensor`, it returns the newly set value correctly. However, when called on subclasses, it incorrectly returns the old value.

## To Reproduce

This prints the correct values `False False True`:

```python
x = Tensor(torch.rand(4,3,16,16))
with torch.no_grad():
    before = x.requires_grad
    y = x.requires_grad_()
    print(before, y.requires_grad, x.requires_grad)
```

This incorrectly prints `False` as the second value:

```python
class FooTensor(Tensor): pass
x = FooTensor(torch.rand(4,3,16,16))
with torch.no_grad():
    before = x.requires_grad
    y = x.requires_grad_()
    print(before, y.requires_grad, x.requires_grad)
```

## Expected behavior

`False False True` should be printed.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 7.0.0 (https://github.com/apple/swift-clang.git 0e8e8c3f4790ab55cacb0388571fecbdd2627e6b)85bad068f01)
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX
GPU 2: TITAN RTX
GPU 3: TITAN RTX
GPU 4: TITAN RTX
GPU 5: TITAN RTX
GPU 6: TITAN RTX

Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.4
/usr/local/cuda-10.0/lib64/libcudnn.so.7.6.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-ignite==0.4.2
[pip3] pytorch-lightning==1.1.3
[pip3] torch==1.7.1
[pip3] torchvision==0.8.2
[conda] blas                      2.16                        mkl    conda-forge
[conda] cudatoolkit               11.0.221             h6bb024c_0    fastchan
[conda] libblas                   3.8.0                    16_mkl    conda-forge
[conda] libcblas                  3.8.0                    16_mkl    conda-forge
[conda] liblapack                 3.8.0                    16_mkl    conda-forge
[conda] liblapacke                3.8.0                    16_mkl    conda-forge
[conda] mkl                       2020.2                      256
[conda] numpy                     1.19.4           py38h18fd61f_2    conda-forge
[conda] pytorch                   1.7.1           py3.8_cuda11.0.221_cudnn8.0.5_0    fastchan
[conda] pytorch-ignite            0.4.2                    pypi_0    pypi
[conda] pytorch-lightning         1.1.3                    pypi_0    pypi
[conda] torchvision               0.8.2                py38_cu110    fastchan
```

**Comment by jph00** [comment]
cc @hameerabbasi

**Comment by hameerabbasi** [comment]
The second value in the `Tensor` case for me is `True`, and in the `FooTensor` case it's `False`. I believe that's what you meant?

**Comment by hameerabbasi** [comment]
I can come up with a smaller reproducer:
Here's a logically smaller reproducer, even if it's bigger in code.

```python
import torch
from torch import Tensor

class FooTensor(Tensor): pass
x = FooTensor(torch.rand(4,3,16,16))
with torch.no_grad():
    before = x.requires_grad
    with torch._C.DisableTorchFunction():
        y = x.requires_grad_().as_subclass(FooTensor)
    print(before, y.requires_grad, x.requires_grad)
```

Strange thing is, it's an interaction between `as_subclass` and `no_grad`. Inside a `no_grad` context, `as_subclass` doesn't do the right thing (it doesn't copy over `requires_grad`, instead setting it to `False`). As `as_subclass` is to be an exact view, this is a bug. Thoughts, @ezyang?

**Comment by jph00** [comment]
> The second value in the `Tensor` case for me is `True`, and in the `FooTensor` case it's `False`. I believe that's what you meant?

Yes - sorry!

**Comment by hameerabbasi** [comment]
An even smaller reproducer:

```python
import torch
from torch import Tensor

class FooTensor(Tensor): pass
x = Tensor(torch.rand(4,3,16,16))
with torch.no_grad():
    before = x.requires_grad
    y = x.requires_grad_().as_subclass(FooTensor)
    print(before, y.requires_grad, x.requires_grad)
```

**Comment by albanD** [comment]
Does the tensor subclassing has special logic to handle the requires_grad field?

**Comment by hameerabbasi** [comment]
The last example highlights the issue rather nicely. `Tensor.as_subclass` uses `Tensor.alias` in C++, which I'm assuming drops the gradient from the graph in a `no_grad` context.

**Comment by albanD** [comment]
Yes.
In general, it is expected that any op that returns a new Tensor within a no grad bloc will generate a Tensor that does not require gradients.
So I could argue that the last example is "ok" as I would expect the `.as_subclass()` to return a different Tensor and thus one that does not require gradients.

But it is indeed surprising in the original example where this is done implicitly.

Also note that for regular Tensors, the current behavior of `.requires_grad` being True in no_grad blocks is not a hard choice that we made. And we might want to change that in the future.

**Comment by hameerabbasi** [comment]
> Also note that for regular Tensors, the current behavior of `.requires_grad` being True in no_grad blocks is not a hard choice that we made. And we might want to change that in the future.

Does this imply that this bug is invalid and should be closed as "wontfix"? I think it's a real bug, but I don't have a proposed fix. Perhaps use the reverse of the `no_grad` block while creating the `.alias`? Is there something that'll temporarily disable the effects of ALL context managers, as I could see variants of this bug coming up in other cases.

**Comment by albanD** [comment]
I don't think it is invalid. But I am not sure we should make a decision on this one until https://github.com/pytorch/pytorch/issues/11390 is decided.
